_posts/function-contracts.md

```diff
@@ -0,0 +1,298 @@
+
+## Dangers of Stubbing
+
+Let us illustrate this with an example. Consider Euclid’s formula for calculating the greatest common divisor of two numbers, here called `max` and `min`. For verification this would be considered an expensive computation due to the recursive call. This recursion is very busy, in the [worst case](https://en.wikipedia.org/wiki/Euclidean_algorithm#Worst-case) the number of steps (recursions) approaches 1.5 times the number of bits needed to
+represent the number. Meaning that for two large 32 bit numbers it can take almost 48 iterations for a single call to `gcd`. If a harness that uses `gcd` does not heavily constrain the input, Kani would have to unroll the recursion 48 times and then execute it symbolically. An expensive operation.
+
+```rust
+fn gcd(mut max: i32, mut min: i32) -> i32 {
+    if min > max {
+				std::mem::swap(&mut max, &mut min);
+    }
+
+    let rest = max % min;
+    if rest == 0 { min } else { gcd(min, rest) }
+}
+
+```
+
+To verify efficiently we would employ a stub for this function that uses assumptions and non-deterministic values to avoid a loop. Our first attempt at stubbing may be very simply that the return value of `gcd` is smaller than its inputs.
+
+```rust
+fn gcd_stub(max: i32, min: i32) -> i32 {
+    let result = kani::any();
+    kani::assume(result < max && result < min);
+    result
+}
+
+```
+
+Now you may immediately notice that this stub is not correct if `max == min`, in which case `gcd` would return `min`, whereas `gcd_stub` would return a non-deterministic value smaller than `min`. If this is used in a harness it can miss a classic off-by-one error.
+
+```rust
+#[kani::proof]
+#[kani::stub(gcd, gcd_stub)]
+fn main() {
+    let x = kani::any();
+    kani::assume(x <= 100);
+    let y = kani::any();
+
+    let v : Vec<i32> = (0..100).collect();
+
+    v[gcd(x, y)]
+}
+
+```
+
+This harness will succeed *despite* the fact that for `x == 100` we are
+accessing out of bounds of the vector. To guard against this problem we can
+check that the assumptions we make in the stub hold for the actual result of
+running `gcd`. This means we ensure that the stub is an overapproximation. This
+harness mirrors our stub but flips the assumption to an assertion. Checking the
+harness will fail and alert us that we would have to change our stub to at least
+`result <= max` and `result <= min` to be *sound*.
+
+```rust
+#[kani::proof]
+fn gcd_stub_check() -> i32 {
+    let max = kani::any();
+    let min = kani::any();
+    let result = gcd(max, min);
+    kani::assert(result < max && result < min && result != 0);
+}
+
+```
+
+## Checking Contracts
+
+What we just did is manually created a function contract. We used the condition `result <= max && result <= min && result != 0` both as an approximation of `gcd` and
```

Comment:
I got slightly confused here. What we generated was a verification harness that checks a post condition, not to be confused with the contract itself.